### DISCUSSION TITLE
What is `mode="python"` for and why is it the default?

### DISCUSSION THREAD

**Discussion by mac01021** [pydantic/pydantic#8886] [Question] [answered]
We use Pydantic for domain modeling and to serialize and deserialize our domain objects to and from json (the form in which they are stored in the database). 

With the introduction of a new domain class to our codebase, I have run into [this issue](https://github.com/pydantic/pydantic/issues/4186#issuecomment-1962420582 ) that prevents the dumping of a model containing a set of frozen models.  As of Pydantic v2, I can call, `thing.model_dump(mode="json")` to dump `thing` successfully, which works fine for my purposes.

But it leaves me wondering what the default `mode="python"` is _for_ and why it is the default.
* `mode="json"` has a clear purpose.  All output consists entirely of json-compatible objects (strings, dicts, lists, numbers).  Models map to dicts.  Sets map to lists. Dates and uuids map to strings.
* `mode=python` is more mysterious.  It makes no particular guarantees AFAICT about what types of python objects may appear in the output. Models map to dicts still, but sets map to sets rather than lists.  What is the use case for transforming all the models into dicts but not transforming other types of things?  Why do we explicitly want sets to map to sets?  What other differences exist between python mode and json mode?

**Comment by samuelcolvin** [comment, marked as answer]
Well it's for any case where you want to create a dict from a model, optionally with filtering and removal of defaults or `None`.

I think I'm general people find it pretty useful.

It's behavior is fairly well defined, it leaves most values unchanged but recursively converts models and dataclassea to dicta.

Python mode is the default, because JSON model was added in V2 and changing the default would have been smoother unnecessary breaking change.

Does that make sense?

**Reply by mac01021** [reply]
Certainly backward compatibility is an excellent reason for python mode to be the default, if python mode behaves the same as `dict` in V1.

I still can't readily think of a common, specific case of "any case where you want to create a dict from a model, optionally with filtering and removal of defaults".  If json mode had been the V1 behavior, would there still be a python mode in V2?  In what cases, if any, would it be preferred?
